crypto, refactor: add new KeyPair class

josibake opened the pull request:
Broken out from #28201 

---

The wallet returns an untweaked internal key for taproot outputs. If the output commits to a tree of scripts, this key needs to be tweaked with the merkle root. Even if the output does not commit to a tree of scripts, BIP341/342 recommend commiting to a hash of the public key.

Previously, this logic for applying the taptweak was implemented in the `CKey::SignSchnorr` method.

This PR moves introduces a KeyPair class which wraps a `secp256k1_keypair` type and refactors SignSchnorr to use this new KeyPair. The KeyPair class is created with an optional merkle_root argument and the logic from BIP341 is applied depending on the state of the merkle_root argument.

The motivation for this refactor is to be able to use the tap tweak logic outside of signing, e.g. in silent payments when retrieving the private key (see #28201). 

Outside of silent payments, since we almost always convert a `CKey` to a `secp256k1_keypair` when doing anything with taproot keys, it seems generally useful to have a way to model this type in our code base.

theuni reviewed (commented):
This function feels kinda weird. As it's named `ApplyTapTweak`, I would expect it to modify `*this`. It also takes in a `uint256* merkle_root` but doesn't check for null, a reference would make more sense.

Perhaps this would make more sense as a static utility function that takes input/output keys?

josibake commented:
> This function feels kinda weird. As it's named `ApplyTapTweak`, I would expect it to modify `*this`. It also takes in a `uint256* merkle_root` but doesn't check for null, a reference would make more sense.
> 
> Perhaps this would make more sense as a static utility function that takes input/output keys?

Utility function might make more sense here: could do a utility function with `secp256k1_keypairs` as in/out args:

```
int compute_taptweak(secp256k1_keypair in, unsigned char merkle_root, secp256k1_keypair out)
```

and use that inside `::SignSchnorr`. This means we still only creating one `secp256k1_keypair` when signing. For usage outside of signing, wrap that utility function in a function takes and returns `CKey`s as arguments:

```
CKey tweaked_key ComputeTapTweak(CKey& internal_key, uint256& merkle_root) {
    // .. do stuff
    compute_taptweak(..);
    return tweaked_key;
```

WDYT?

theuni commented:
`ComputeTapTweak` is more like what I had in mind, yes.

josibake commented:
Updated to use a utility function (instead of a method on `CKey`) per the @theuni 's feedback

theuni commented on src/key.cpp:
```diff
@@ -275,20 +275,51 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
-        if (!secp256k1_xonly_pubkey_serialize(secp256k1_context_sign, pubkey_bytes, &pubkey)) return false;
-        uint256 tweak = XOnlyPubKey(pubkey_bytes).ComputeTapTweakHash(merkle_root->IsNull() ? nullptr : merkle_root);
-        if (!secp256k1_keypair_xonly_tweak_add(secp256k1_context_static, &keypair, tweak.data())) return false;
+        CKey tweaked_key;
```
Will anything ever use this key directly? Or will it always just be turned back into a `secp256k1_keypair`? It seems the CKey output is just a wrapper that causes overhead here. I guess for the sake of not exposing `secp256k1_keypair` to the caller?

josibake replied:
Yep! From the description:

> The motivation is to be able to use the ::ApplyTapTweak logic outside of signing, e.g. in silent payments when retrieving the private key. Outside of silent payments, having this method would support any future use cases where the tweaked key is needed outside of signing.

This commit was broken out of the silent payments sending PR, because there we need the tweaked `CKey`. However, we do eventually turn it back into a `keypair` when passing it to libsecp, so maybe there is an argument for just returning the `keypair`? But passing around libsecp types seemed a bit weird..

theuni replied:
Hmm, if what you need is a wrapped `keypair`, how about creating a wrapped `keypair`? :) https://github.com/theuni/bitcoin/commit/1a1599f59f711206ea91f025ee48b148e17b01ce

^^ Completely untested. Maybe state via `ApplyTapTweak` is unnecessary as the ctor could just take a pointer to a merkle tree instead?

josibake replied:
Nice, thanks for the suggestion! This makes a ton more sense. I think it would be better to have the ctor take a pointer to the merkle root because `ApplyTapTweak` is something that you a) only want to do once over the lifetime of the object and b) will always be applied if a `merkle_root` is present (even if its `merkle_root.IsNull() == true`). I don't think this is actually a use case, but if for whatever reason you did need to do something with the key and then later apply a merkle root tweak, you could just use the `CKey` first and then create a `KeyPair(*this, merkle_root)` with the merkle root when needed. 

Also, what do you think about something like `KeyPair CKey::ComputeKeyPair(*merkle_root);` method, instead of `KeyPair(CKey, merkle_root)`?

Will pull this in and test it in #28122 and #28201.

theuni replied:
Please consider the above a doodle, change it however you'd like. I just wanted to sketch out the concept of a wrapped `keypair`.

josibake replied:
Yeah, for sure! But thanks, very helpful doodle. Not sure why I was so apprehensive about passing around a keypair, but seeing you write it out helped it click for me.

theuni replied:
Updated as suggested here: https://github.com/theuni/bitcoin/commits/keypairtaptweak/

josibake replied:
Awesome, thanks @theuni ! I pulled this commit in and also add a "simple read only vector like interface" to the `KeyPair` class (same as `CKey`). Needed for: https://github.com/bitcoin/bitcoin/commit/5af5164b84b534c9a3171d898c09125957b21ba1#diff-e5cbd40011e760f175b82d7ec44ad512bdb3d9108068b660642204da9c9187b1R204

I also rebased #28122 on this PR and confirmed everything works with the new `KeyPair` class.

theuni replied:
That looks much better, thanks. Exposing the vector-like interface is a shame, but I don't see any way around it for `secp256k1_silentpayments_sender_create_outputs`.

theuni commented on src/key.cpp:
```diff
@@ -275,20 +275,51 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
+        CKey tweaked_key;
+        if (!ComputeTapTweak(*this, *merkle_root, tweaked_key)) return false;
+        ret = secp256k1_keypair_create(secp256k1_context_sign, &keypair, UCharCast(tweaked_key.begin()));
+        tweaked_key = CKey();
```
This is going out of scope anyway, no need?

josibake commented:
Updated to use the new `KeyPair` wrapper class (h/t @theuni).

theuni commented on src/key.h:
```diff
@@ -203,6 +205,8 @@ class CKey
                                        const EllSwiftPubKey& our_ellswift,
                                        bool initiating) const;
+
+    KeyPair ComputeKeyPair(const uint256* merkle_root) const;
```
This could use a comment.

josibake replied:
Done.

theuni reviewed (commented):
It's weird to see this hooked up but unused. It could also use some simple test vectors.

Mind killing both birds by adding some tests, at least 1 for each `merkle_root` state?

josibake commented:
@theuni Updated with a comment and added `KeyPair` to the BIP340 test vectors. This does test all possible merkle_root states and ensures everything is 1-to-1 with `XOnlyPubKey::ComputeTapTweakHash` and `CKey::SignSchnorr`

josibake commented:
Updated to add a move constructor to the KeyPair class. I noticed this was missing while trying to use the new class in #28201 (i.e. creating a std::vector of KeyPairs). @theuni assuming this was just missed, but if not curious if you have any objections to adding a move constructor on KeyPair?

theuni reviewed (commented):
Mind changing the dumb c-style casts to `reinterpret_cast` so it's clear that they can't be `static_cast`s ?  Sorry, I know that's my code.

utACK after that.

theuni commented:
PR description needs an update too :)

josibake commented:
@theuni title, description, and dumb c-style casts updated!

theuni reviewed (approved):
ACK bdc2a656c2d2a61d226fde1d1fd4e79664106e18

paplorinc commented on src/test/key_tests.cpp:
```diff
@@ -327,6 +327,16 @@ BOOST_AUTO_TEST_CASE(bip340_test_vectors)
+        // Repeat the same check, but use the KeyPair directly without any merkle tweak
+        KeyPair keypair = key.ComputeKeyPair(/*merkle_root=*/nullptr);
+        CKey keypair_seckey;
+        BOOST_CHECK(keypair.GetKey(keypair_seckey));
```
nit: BOOST_CHECK_MESSAGE often produces more readable results, but it may be inconvenient to set it up properly. If you think it adds value, it may be helpful to add some debug info as a message in case of likely failures.

paplorinc commented on src/key.h:
```diff
@@ -236,6 +252,41 @@ struct CExtKey {
+private:
+    KeyPair(const CKey& key, const uint256* merkle_root);
+
+    using KeyType = std::array<unsigned char, 96>;
```
nit: would it make sense to extract the `96` to a const?
```C++
constexpr size_t SECP256K1_KEYPAIR_SIZE = 96;
using KeyType = std::array<unsigned char, SECP256K1_KEYPAIR_SIZE>;
```

josibake replied:
I don't really see a benefit. We shouldn't be exposing this value to anything outside the class, so defining a constexpr here seems unnecessarily verbose.

itornaza reviewed (approved):
tested ACK bdc2a656c2d2a61d226fde1d1fd4e79664106e18

Built from source and run all unit, functional and extended tests and all of them pass. Looked through the code changes and they look great to me. I particularly like the default definition of the move constructors in `KeyPair` since there is a pointer member `m_keydata` in that class, plus the fact that the generation of the copy constructors that are not needed anywhere is suppressed due to the use of the default keyword which I consider an excellent programming practice. One small suggestion about repeating the KeyPair header for your consideration if you ever need to update for other more serious reasons.

paplorinc commented on src/key.cpp:
```diff
@@ -277,27 +277,7 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
 
 bool CKey::SignSchnorr(const uint256& hash, Span<unsigned char> sig, const uint256* merkle_root, const uint256& aux) const
 {
-    assert(sig.size() == 64);
```
Moving and modifying the code in the same commit (while many of the changes serve different purposes, mixing low and high-risk changes) makes the review a lot harder since the diff isn't really useful – having separate red and green parts, not really a diff.

Additionally, the commit doesn't explain each change separately. For example:
* The exact reason for memory management changes (via `make_secure_unique`) is not explained, nor are the security implications or potential performance impacts discussed.
* It's unclear why `GetKey` is only used in tests. Why not delete it and introduce it in another PR where it's actually needed by production code?
* The fate of `memory_cleanse(&keypair, sizeof(keypair))` is not explained, nor are the implications of its removal or replacement addressed.

It would help a lot if a commit did a single thing, such as a rename or cast change, changing a constant, moving implementation over to a new method, introducing a new class, or splitting a method into two methods, etc.
This would separate trivial changes from high-risk ones that need extra attention, so we don't have unrelated changes competing for our attention.

Please see https://github.com/bitcoin/bitcoin/pull/28280/commits for an example of focused, single-purpose commits that separate low and high-risk changes (e.g., preparatory refactorings from the focus of the PR), making the review a lot easier.

josibake replied:
Hey @paplorinc , thanks for taking a look! I agree, the first commit was quite dense. I've broken the first commit into smaller commits to aid with review and tried to include more information in the commit messages as to the motivation for the changes. Per your feedback, I've also removed the `GetKey` method as its not necessary.

Shuhaib07 replied:
Okay

l0rinc reviewed (changes requested):
Not yet sure about adding stuff that "will hopefully used this way one day".

Refactoring is welcome, as long as it's easy to review, which is not yet the case here, could you split the change into trivial, focused commits?

josibake committed:
tests: add tests for KeyPair

Reuse existing BIP340 tests, as there should be
no behavior change between the two

josibake committed:
crypto: add KeyPair wrapper class

Add a `KeyPair` class which wraps the `secp256k1_keypair`. This keeps
the secret data in secure memory and enables passing the
`KeyPair` object directly to libsecp256k1 functions expecting a
`secp256k1_keypair`.

Motivation: when passing `CKeys` for taproot outputs to libsecp256k1 functions,
the first step is to create a `secp256k1_keypair` data type and use that
instead. This is so the libsecp256k1 function can determine if the key
needs to be negated, e.g., when signing.

This is a bit clunky in that it creates an extra step when using a `CKey`
for a taproot output and also involves copying the secret data into a
temporary object, which the caller must then take care to cleanse. In
addition, the logic for applying the merkle_root tweak currently
only exists in the `SignSchnorr` function.

In a later commit, we will add the merkle_root tweaking logic to this
function, which will make the merkle_root logic reusable outside of
signing by using the `KeyPair` class directly.

Co-authored-by: Cory Fields <cory-nospam-@coryfields.com>

josibake committed:
refactor: move SignSchnorr to KeyPair

Move `SignSchnorr` to `KeyPair`. This makes `CKey::SignSchnorr` now
compute a `KeyPair` object and then call `KeyPair::SignSchorr`. The
notable changes are:

    * Move the merkle_root tweaking out of the sign function and into
      the KeyPair constructor
    * Remove the temporary secp256k1_keypair object and have the
      functions access m_keypair->data() directly

josibake commented:
Updated https://github.com/bitcoin/bitcoin/commit/bdc2a656c2d2a61d226fde1d1fd4e79664106e18 -> https://github.com/bitcoin/bitcoin/commit/5e9f4f41de98c2cdcc74e31fe925b2f5cf9f6aec ([add-apply-taptweak-method-00](https://github.com/josibake/bitcoin/tree/add-apply-taptweak-method-00) -> [add-apply-taptweak-method-01](https://github.com/josibake/bitcoin/tree/add-apply-taptweak-method-01) ([compare](https://github.com/josibake/bitcoin/compare/add-apply-taptweak-method-00..josibake:add-apply-taptweak-method-01))

* Follows some `CKey` conventions with `MakeKeyPairData`, `ClearKeyPairData`, etc
* Breaks up the first commit into smaller, more focused commits to aid review (h/t @paplorinc)

Not much in terms of logical changes from the previous version, but broken into incremental steps to aid with review. Worth mentioning for reviewers: `[staging]` commits are changes that should not be considered final on their own but are incremental steps to make the overall change easier to review. Each commit is still atomic in that it compiles on its own, but having them in smaller steps like this helps quite a bit in understanding the diffs. However, if reviewers feel this is unnecessary, I'm happy to squash these staging commits back into a single commit.

This push also includes a rebase on master, but the [compare](https://github.com/josibake/bitcoin/compare/add-apply-taptweak-method-00..josibake:add-apply-taptweak-method-01) is provided with both the old and new branch rebased on master.

josibake commented:
cc @theuni - would appreciate your eyes on this again

josibake commented on src/key.h:
```diff
@@ -236,6 +252,41 @@ struct CExtKey {
     void SetSeed(Span<const std::byte> seed);
 };
 
+/** KeyPair
```
Hey @itornaza , thanks for the review! I've incorporated your feedback and moved the comment regarding the merkle root to the `ComputeKeyPair` function and made this comment more specific to the class itself.

paplorinc commented on src/key.cpp:
```diff
@@ -272,21 +272,23 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
         unsigned char pubkey_bytes[32];
-        if (!secp256k1_xonly_pubkey_serialize(secp256k1_context_sign, pubkey_bytes, &pubkey)) return false;
+        ret &= secp256k1_xonly_pubkey_serialize(secp256k1_context_sign, pubkey_bytes, &pubkey);
         uint256 tweak = XOnlyPubKey(pubkey_bytes).ComputeTapTweakHash(merkle_root->IsNull() ? nullptr : merkle_root);
```
off topic:  we might need a [copy of pubkey_bytes](https://github.com/bitcoin/bitcoin/blob/b946f8a4c51be42e52d63a6d578158c0b2a6b7ed/src/pubkey.cpp#L199) in this case, but I guess that's unrelated to the PR

paplorinc commented on src/key.cpp:
```diff
@@ -272,21 +272,23 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
+        ret &= secp256k1_keypair_xonly_pub(secp256k1_context_static, &pubkey_verify, nullptr, &keypair);
         ret &= secp256k1_schnorrsig_verify(secp256k1_context_static, sig.data(), hash.begin(), 32, &pubkey_verify);
     }
     if (!ret) memory_cleanse(sig.data(), sig.size());
```
off topic: I'm out of my depth here, but curious, what's the reason for not needing a cleanup of keypair in the merkle_root branch as well?

josibake replied:
Good catch, I think you're right that if `xonly_tweak_add` were to fail (e.g. tweak32 is the negation of the secret key), then we would return without explicitly clearing the keypair variable. That being said, the variable should already be cleared when it goes out of scope, so the explicit `memory_cleanse` is more of a belt and suspenders approach, from what I understand.

paplorinc commented on src/key.cpp:
nit: typos in commit message: `Accumlute` and `returning early on the function`

josibake replied:
Fixed.

l0rinc reviewed (commented):
Started reviewing, but got a bit stuck at the first commit, will continue a bit later

paplorinc commented on src/key.h:
```diff
@@ -207,8 +207,19 @@ class CKey
+     *                               used for signatures in BIP342 script).
+     *  - If merkle_root->IsNull():  tweak the internal key with H_TapTweak(pubkey) (this is used for
+     *                               key path spending when no scripts are present).
+     *  - Otherwise:                 tweak the internal key H_TapTweak(pubkey || *merkle_root)
```
```suggestion
     *  - Otherwise:                 tweak the internal key with H_TapTweak(pubkey || *merkle_root)
```

josibake replied:
Fixed.

paplorinc commented on src/key.h:
```diff
@@ -207,8 +207,19 @@ class CKey
+     *  - If merkle_root->IsNull():  tweak the internal key with H_TapTweak(pubkey) (this is used for
+     *                               key path spending when no scripts are present).
+     *  - Otherwise:                 tweak the internal key H_TapTweak(pubkey || *merkle_root)
+     *                               (this is used for key path spending, with specific
```
```suggestion
     *                               (this is used for key path spending with a specific
```

josibake replied:
Fixed.

paplorinc commented on src/key.h:
```diff
@@ -271,20 +285,27 @@ class KeyPair
-     * This method will be made read-only in a later commit.
-     * It is only read/write here to simplify the diff in the next two commits
+     * This is provided as a read-only method for passing a KeyPair object to secp256k1 functions
+     * expecting a `secp256k1_keypair`. This avoids need to create a `secp256k1_keypair` object and
```
```suggestion
     * expecting a `secp256k1_keypair`. This avoids the need to create a `secp256k1_keypair` object and
 ```

josibake replied:
Fixed.

paplorinc commented on src/key.cpp:
nit: typo in commit message: `Instead of calling ... instead invalidate`

josibake replied:
Fixed.

paplorinc commented on src/key.cpp:
```diff
@@ -271,19 +271,11 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
-    memory_cleanse(&keypair, sizeof(keypair));
+    bool ret = keypair.SignSchnorr(hash, sig, aux);
+    /* Clear the keypair */
+    keypair = KeyPair();
```
Might be a naive question, but is this a behavior change, does it have any security implications (e.g. could it leave keys in the memory or something)?

itornaza replied:
I am also wandering how the assignment involving the default constructor will clear the existing keypair data from memory, but I might be naive as well!

josibake replied:
> Since keypair is a local variable in this function shouldn't the KeyPair destructor do the secure erase job anyway at the end of the scope through RAII since m_keypair is a secure_unique_ptr

Correct, which should happen regardless of whether or not we explicitly clear the key data with `keypair = KeyPair();`. Having `keypair = KeyPair()` a belt and suspenders by explicitly clearing the variable. This is because if we create a `KeyPair` object with the default constructor, there is no call to `MakeKeyPairData`, hence `m_keypair` is empty. When we assign this default constructed `KeyPair` to `keypair`, it calls `ClearKeyPairData` on `keypair`, which resets it. The assignment operator was copied from `CKey` and you can see an example of the same pattern used here: https://github.com/bitcoin/bitcoin/blob/5e9f4f41de98c2cdcc74e31fe925b2f5cf9f6aec/src/bip324.cpp#L66-L70

paplorinc replied:
I'm still a bit hesitant here, see my related comment

josibake replied:
My last comment was wrong, @itornaza is correct: the KeyPair destructor is doing the secure erase job and the `kp = KeyPair();` is unnecessary.

l0rinc reviewed (commented):
Very nice separation into commits, left some questions

itornaza commented on src/key.cpp:
```diff
@@ -271,26 +271,11 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
-    if (!ret) memory_cleanse(sig.data(), sig.size());
-    memory_cleanse(&keypair, sizeof(keypair));
+    KeyPair keypair = ComputeKeyPair(merkle_root);
+    if (!keypair.IsValid()) return false;
```
It would be trivial to also erase the `keypair` even if it turns out not to be valid, or would that be paranoid?

josibake replied:
`IsValid()` returns false when `m_keypair` data is a nullptr, which effectively means the key data was never created or has been cleared (see `src/support/allocators/secure.h`)

itornaza reviewed (approved):
re tACK 5e9f4f41de98c2cdcc74e31fe925b2f5cf9f6aec

Thank you for your commit redo in a more granular way. Just a couple of questions really in the discussion above. To be thorough, rebuilt and retested the commit and all tests included the extended suite pass.

josibake commented on src/key.cpp:
```diff
@@ -272,21 +272,23 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
+        ret &= secp256k1_keypair_xonly_tweak_add(secp256k1_context_static, &keypair, tweak.data());
     }
-    bool ret = secp256k1_schnorrsig_sign32(secp256k1_context_sign, sig.data(), hash.data(), &keypair, aux.data());
+    if (!ret) return false;
```
I'm fairly confident this is not a behavior change, but could have missed something. 

If you look at `secp256k1_keypair_xonly_pub`, this will only fail if there was an issue creating the `keypair` object, but this is checked above so we would have already short-circuited. The same is true for `secp256k1_keypair_xonly_pubkey_serialize` (only fails if we can't load the `keypair`, which only fails if the `keypair` is malformed).

`secp256k1_keypair_xonly_tweak_add` will fail if `tweak32` is greater than the curve order or if the tweak is the negation of the secret key.

Putting it all together, if we can't create a valid `keypair`, the function will exit (same as before). If we do create a valid `keypair`, then the only error path we can hit is `secp256k1_keypair_xonly_tweak_add` and if we do hit an error here, we exit before attempting to sign (same as before).

paplorinc replied:
Your version always executes all steps (e.g. computes `XOnlyPubKey(pubkey_bytes).ComputeTapTweakHash(merkle_root->IsNull() ? nullptr : merkle_root);` even after `secp256k1_keypair_load` wasn't successful, right? Previously this method wasn't executed, so at best it's a performance difference, at worst we're hoping that each method handles this error properly (which wasn't necessary before, since the first method guarded them).

josibake replied:
Right, my point is there is no way to call `CKey::SignSchnorr` such that `_keypair_create` will succeed but `_xonly_pub` or `_serialize` will fail. Annoyingly, this means in the old version we have branches in the code that we cannot test (easy to prove me wrong if you have a test case :wink:). Hence, I stand by the statement this is not a behavior change and we are merely disagreeing over code aesthetics.

I think you make a good point that _future_ changes to libsecp256k1 could change this (although, I'd argue if we can create a keypair object but cannot extract and serialize the pubkey portion something is seriously broken), so I can add back the early returns. However, I'll reiterate that I find it to be a bit smelly when we have if branches in the code we cannot test.

paplorinc replied:
Do you agree that computations are done now, that would have been avoided previously? It might return the same value, but it does different calculations than before.

> it to be a bit smelly when we have if branches in the code we cannot test.

absolutely, but I don't yet see how avoiding short-circuiting helps with the testing - but exploding it into multiple methods certainly did!

josibake replied:
> Do you agree that computations are done now, that would have been avoided previously?

No, this is the part I am disagreeing with. As I explained in the first message, there is no way for `_xonly_pub` or `_serialize` to fail if `keypair_create` succeeds. If `_create`, `_xonly_pub`, and `_serialize` succeed, then `XOnlyPubKey(pubkey_bytes).ComputeTapTweakHash(merkle_root->IsNull() ? nullptr : merkle_root);` is computed. After that, `_xonly_tweak_add` will be called. `_xonly_tweak_add` can fail if tweak32 is greater than the curve order or if it is the negation of the private key that was used to create `keypair`. If this function fails, `CKey::SignSchnorr` will exit before attempting to sign, which is the same behavior in the old version and the new version.

> absolutely, but I don't yet see how avoiding short-circuiting helps with the testing

It helps by removing dead branches from the code. I'd be much happier to keep it broken out into multiple if branches if we had test cases to exercise the if branches. But if we have if branches that can realistically never be hit, I think it makes the code overly complicated. In this case, we end up with a more complicated function in the hopes of preventing an unnecessary call to `ComputeTapTweakHash`, which realistically will never happen and hardly seems worth it given this is a relatively cheap operation (and one without side effects).

paplorinc replied:
> there is no way for _xonly_pub or _serialize to fail if keypair_create succeeds

I wish it were easy to see that from the code, like before the change.
I'll check it out more thoroughly tomorrow - if you think you can make it more obvious in the code (e.g. if you're sure some of the values can only be true, you could maybe signal that by replacing confusing `ret &= ` code with asserts instead), please push it.

josibake replied:
Spent some time thinking about this and realised the main issue here is we start with a valid secret key, create a keypair (which generates the public key in the process), and then extract the public key from the keypair and serialise it. Instead, we can just get the public key directly from the secret key via `CKey::GetPubKey`. This makes the original code more clear, eliminates the unnecessary if branches and makes the diff _much_ simpler.

If you look into `GetPubKey` you can see it’s calling the same libsecp256k1 functions to generate the public key, but asserting true for the serialisation and removes the unnecessary load steps.

paplorinc replied:
beautiful!

josibake replied:
Yep, this will be slower (in cases where we don't have a merkle root and instead use the public key per BIP341's recommendation). I looked to see if we had a benchmark for signing to see if I could quantify this, but didn't see one. Was thinking about writing one but then talked myself out of it because I don't think performance is as much a concern in signing (as opposed to verification, where it absolutely is a concern).

In this case, I do think improving the clarity of the code and removing extra steps at the cost of performance seems worth it? I can also write a benchmark to quantify if performance here is indeed a concern.

josibake replied:
Ah, nice! I'll see if I can easily modify the benchmark to validate that the difference is indeed negligible (and quantify the difference in the event it is not negligible).

> Just wanted to point out that this could be a slight drawback.

It's a good observation! I have a slight preference for clarity over performance in this context, but good you mentioned it in case others feel differently.

paplorinc replied:
@josibake, what do you think of this?

josibake replied:
Thanks for writing the benchmark @paplorinc ! I've added a _slightly_ modified version in the first commit. 15% is a tough pill to swallow, so I reverted back to extracting the xonly public key from the secp256k1_keypair object and serialising it. Per our early conversation, I replaced the early returns with asserts. This more closely mirrors what `CKey::GetPubKey()` is doing and I think better documents what the code is doing.

I added this as the last commit to keep the diffs as clean as possible.

paplorinc replied:
I think you found a good compromise, especially with the updated test which checks that it basically does the same thing.

josibake commented on src/key.cpp:
```diff
@@ -272,21 +272,23 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
     if (merkle_root) {
         secp256k1_xonly_pubkey pubkey;
-        if (!secp256k1_keypair_xonly_pub(secp256k1_context_sign, &pubkey, nullptr, &keypair)) return false;
+        ret &= secp256k1_keypair_xonly_pub(secp256k1_context_sign, &pubkey, nullptr, &keypair);
```
Personally, I don't see any reason to not reuse `ret` and I think it makes the logic easier to follow vs re-initializing the variable. Regarding returning inside the `if (ret)`, as mentioned in https://github.com/bitcoin/bitcoin/pull/30051#discussion_r1685446875, `keypair_xonly_pub` will only fail if the `keypair` is malformed, so if we've successfully created the keypair this call should never fail. So if `schnorrsig_verify` fails, we check it on the next line and clear the signature data before returning. I find this easier to follow than having a return inside the `if (ret)` block (and we don't gain anything by having an early return inside the `if (ret)` block.

josibake replied:
Resolving since this is no longer relevant.

josibake commented on src/key.cpp:
```diff
@@ -273,22 +273,22 @@ bool CKey::SignSchnorr(const uint256& hash, Span<unsigned char> sig, const uint2
-    secp256k1_keypair keypair;
-    if (!secp256k1_keypair_create(secp256k1_context_sign, &keypair, UCharCast(begin()))) return false;
+    KeyPair keypair = ComputeKeyPair();
+    if (!keypair.IsValid()) return false;
```
Hm, I don't think it makes much of a difference for the diff and makes this particular staging commit more confusing by creating the `KeyPair` object just to pull out the data back out into a `secp256k1_keypair` object. The main advantage of the `KeyPair` wrapper class is that we are keeping the keypair in secure memory, which we would be undoing here by copying it out into a `keypair` object.

paplorinc replied:
there's a lot of ugly duplication because of all the `reinterpret_cast` changes. Wouldn't it make sense to extract those and simplify the function and the diff?

paplorinc replied:
Or maybe introduce a casting helper like https://github.com/bitcoin/bitcoin/blob/master/src/span.h#L295

josibake replied:
Ah, sorry, I misunderstood your suggestion as "extract" (copy) the data out into a variable but you're suggesting saving the pointer as a variable. I think this would simplify the diff a lot, will give it a shot.

The only thing I don't like here is needing to pass it as non-const to two functions and const to the other two, but I think its okay to pass a non-const object to a function with const in the signature?

paplorinc replied:
if you end up using similar casting helper methods, the const part won't matter anymore, right?

paplorinc replied:
nice!

josibake commented on src/key.cpp:
```diff
@@ -273,22 +273,22 @@ bool CKey::SignSchnorr(const uint256& hash, Span<unsigned char> sig, const uint2
         ret &= secp256k1_xonly_pubkey_serialize(secp256k1_context_sign, pubkey_bytes, &pubkey);
         uint256 tweak = XOnlyPubKey(pubkey_bytes).ComputeTapTweakHash(merkle_root->IsNull() ? nullptr : merkle_root);
-        ret &= secp256k1_keypair_xonly_tweak_add(secp256k1_context_static, &keypair, tweak.data());
+        ret &= secp256k1_keypair_xonly_tweak_add(secp256k1_context_static, reinterpret_cast<secp256k1_keypair*>(keypair.data()), tweak.data());
```
`secp256k1_keypair_xonly_tweak_add` modifies the `keypair` object with `tweak32`, so it can't be const. More generally, libsecp256k1 is a dependency of Bitcoin Core, so we shouldn't ever be modifying `secp256k1_` functions in our code base (despite us pulling it in as a subtree).

paplorinc replied:
right, I thought that `make` would catch that, not sure why it allowed a `memset` on a const (most be C vs C++ boundary, I guess).

josibake replied:
That is surprising. I applied your patch directly in `bitcoin-core/secp256k1` and did get a warning:

```
src/modules/extrakeys/main_impl.h:266:12: warning: passing argument 1 of ‘memset’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  266 |     memset(keypair, 0, sizeof(*keypair));
      |            ^~~~~~~
In file included from src/ecmult_impl.h:10,
                 from src/secp256k1.c:30:
/usr/include/string.h:61:28: note: expected ‘void *’ but argument is of type ‘const secp256k1_keypair *’
   61 | extern void *memset (void *__s, int __c, size_t __n) __THROW __nonnull ((1));
      |                      ~~~~~~^~~
src/modules/extrakeys/main_impl.h:278:32: warning: passing argument 1 of ‘secp256k1_keypair_save’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  278 |         secp256k1_keypair_save(keypair, &sk, &pk);
      |                                ^~~~~~~
src/modules/extrakeys/main_impl.h:156:55: note: expected ‘secp256k1_keypair *’ but argument is of type ‘const secp256k1_keypair *’
  156 | static void secp256k1_keypair_save(secp256k1_keypair *keypair, const secp256k1_scalar *sk, secp256k1_ge *pk) {
```

which is good, but I would have also expected it to fail to compile :thinking:

paplorinc replied:
it must be a C vs C++ inconsistency - we're not affected by it anymore, you can resolve this

josibake commented:
Updated https://github.com/bitcoin/bitcoin/commit/5e9f4f41de98c2cdcc74e31fe925b2f5cf9f6aec -> https://github.com/bitcoin/bitcoin/commit/5d9a6cf6286f9a7f93527ea76b910537d709a860  ([add-apply-taptweak-method-01](https://github.com/josibake/bitcoin/tree/add-apply-taptweak-method-01) -> [apply-taptweak-method-02](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-02) ([compare](https://github.com/josibake/bitcoin/compare/add-apply-taptweak-method-01..josibake:apply-taptweak-method-02))

* Fix typos in commit messages, `src/key.h` (h/t @paplorinc)
* Use `IsValid()` instead of `m_keypair ?` (h/t @paplorinc)

Thanks for the re-review @paplorinc and @itornaza!

josibake commented on src/key.h:
```diff
@@ -235,6 +242,64 @@ struct CExtKey {
+     * This method will be made read-only in a later commit.
+     * It is only read/write here to simplify the diff in the next two commits
+     */
+    unsigned char* data() { return m_keypair ? m_keypair->data() : nullptr; }
```
Agree, I think `IsValid` makes this more readable, updated.

josibake commented on src/key.cpp:
Fixed.

josibake commented on src/key.cpp:
```diff
@@ -272,21 +272,23 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
 bool CKey::SignSchnorr(const uint256& hash, Span<unsigned char> sig, const uint256* merkle_root, const uint256& aux) const
 {
     assert(sig.size() == 64);
+    bool ret = 1;
```
Agree, more readable to have this as `bool ret = true;`, updated.

josibake commented on src/key.cpp:
```diff
@@ -272,21 +272,23 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
     }
-    bool ret = secp256k1_schnorrsig_sign32(secp256k1_context_sign, sig.data(), hash.data(), &keypair, aux.data());
+    if (!ret) return false;
+    ret &= secp256k1_schnorrsig_sign32(secp256k1_context_sign, sig.data(), hash.data(), &keypair, aux.data());
```
Unless there is a functional disadvantage to having it this way, I prefer this for readability as it makes it more clear (to me, anyways) that we are accumulating the return values from the various secp256k1 function calls in the `ret` variable.

paplorinc replied:
it's narrowing the scope, what would be the reason for reusing a variable and widening its scope?
If you feel strongly about it, I don't particularly mind either way.

paplorinc replied:
resolved

paplorinc commented on src/key.cpp:
```diff
@@ -272,21 +272,23 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
 bool CKey::SignSchnorr(const uint256& hash, Span<unsigned char> sig, const uint256* merkle_root, const uint256& aux) const
 {
     assert(sig.size() == 64);
+    bool ret = true;
```
nit: we could move this down, after the first early return, since it's not needed before that

josibake replied:
Removed all of the instances of initialisation the bool outside of a function call.

l0rinc reviewed (changes requested):
Thanks for the fixes!
I'm still not convinced that the lack of short-circuits didn't introduce any change.
Please prove me wrong, I want to ACK this :)

josibake commented:
Update https://github.com/bitcoin/bitcoin/commit/5d9a6cf6286f9a7f93527ea76b910537d709a860 ->  https://github.com/bitcoin/bitcoin/commit/49057fc4f1e73a14f673c934573d727ae0229779 ([apply-taptweak-method-02](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-02) -> [apply-taptweak-method-03](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-03) ([compare](https://github.com/josibake/bitcoin/compare/apply-taptweak-method-02..josibake:apply-taptweak-method-03))

* Use `CKey::GetPubKey()` instead of extracting and serializing the public key from the keypair object
* Save `reinterpret_cast` as a variable to simplify diff/code (h/t @paplorinc )

Overall, I think this simplifies things quite a bit. If you look at the internals of `CKey::GetPubKey()`, there are already `asserts` for intermediate libsecp256k1 function calls which make it clear these calls (`_serialize`, for example) are always expected to succeed when using a valid secret key.

paplorinc commented on src/key.cpp:
```diff
@@ -433,6 +424,21 @@ KeyPair::KeyPair(const CKey& key, const uint256* merkle_root)
 
+bool KeyPair::SignSchnorr(const uint256& hash, Span<unsigned char> sig, const uint256& aux) const
+{
+    assert(sig.size() == 64);
```
nit: it's 64 because of https://github.com/bitcoin/bitcoin/blob/master/src/script/sign.cpp#L86, right?
It's not really called from other places, the assert seems like noise here.

josibake replied:
It's 64 because Schnorr signatures, unlike ECDSA signatures, are by definition 64 bytes (see https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#description), so we definitely want to keep this assert.

paplorinc replied:
Yes, of course, but that's already the case for every caller currently (i.e. not part of a public api, we know all the callers), and we're not really asserting this in other such methods, so it seems redundant to me. If you disagree, just resolve it, not a big deal either way.

paplorinc commented on src/test/key_tests.cpp:
```diff
@@ -299,6 +300,13 @@ BOOST_AUTO_TEST_CASE(bip340_test_vectors)
+        bool kp_ok = keypair.SignSchnorr(msg256, sig64, aux256);
+        BOOST_CHECK(kp_ok);
+        BOOST_CHECK(pubkey.VerifySchnorr(msg256, sig64));
+        BOOST_CHECK(std::vector<unsigned char>(sig64, sig64 + 64) == sig);
```
`BOOST_CHECK` has very bad error messages, consider using the `_EQUAL` or in this case `_EQUAL_COLLECTIONS` for better errors:
```suggestion
        BOOST_CHECK_EQUAL_COLLECTIONS(std::begin(sig64), std::end(sig64), sig.begin(), sig.end());
```

josibake replied:
In this case, I think it's better to match the style of the existing test but I've made a note regarding `_EQUAL` and `_EQUAL_COLLECTIONS` going forward.

paplorinc commented on src/key.cpp:
```diff
@@ -271,26 +271,11 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
+    if (!kp.IsValid()) return false;
+    bool ret = kp.SignSchnorr(hash, sig, aux);
+    /* Clear the keypair */
+    kp = KeyPair();
```
this is the only remaining part that I'm uncomfortable with, even the IDE is complaining:
<img  src="https://github.com/user-attachments/assets/52bf7173-a562-4dde-9f2c-9834b3a5d5da">

which I'm not getting for `BIP324Cipher::Initialize`, since it's assigning a field, not a local variable.

Is there a more intuitive way to do this? Do we have tests to check that this value is cleared?

josibake replied:
Ah! I was misunderstanding `m_key = CKey()`. Since CKey (and KeyPair) are using secure allocator, the contents will be cleared before deletion (e.g. when the variable goes out of scope). But in the case of `BIP324Cipher::Initialize`, it seems `m_key` is used to initialize the connection, but then needs to be explicitly cleared by setting with an empty CKey (invalid CKey), because the member field will not go out of scope and be cleared.

So in this case, we don't need the `kp = KeyPair();` line at all. Also worth mentioning: `memory_cleanse(keypair..` was needed before because the keypair object was _not_ using a secure allocator.

l0rinc reviewed (commented):
Nice, left a few nits and a clarification

josibake committed:
tests: add key tweak smoke test

Sanity check that using CKey/CPubKey directly vs using secp256k1_keypair objects
returns the same results for BIP341 key tweaking.

Co-authored-by: l0rinc <pap.lorinc@gmail.com>

josibake commented:
Update https://github.com/bitcoin/bitcoin/commit/49057fc4f1e73a14f673c934573d727ae0229779 - > https://github.com/bitcoin/bitcoin/commit/b8b3a9f18670ec7bf246a57950cdae7e034a264d([apply-taptweak-method-03](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-03) -> [apply-taptweak-method-04](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-04) ([compare](https://github.com/josibake/bitcoin/compare/apply-taptweak-method-03..josibake:apply-taptweak-method-04))

* Renamed `ret` to `success` in KeyPair constructor (h/t @paplorinc)
* Removed unnecessary `KeyPair()` call
* Added key tweak smoke test (h/t @paplorinc )

josibake commented on src/key.cpp:
```diff
@@ -422,13 +419,17 @@ void CExtKey::Decode(const unsigned char code[BIP32_EXTKEY_SIZE]) {
     MakeKeyPairData();
     auto keypair = reinterpret_cast<secp256k1_keypair*>(m_keypair->data());
-
     bool ret = secp256k1_keypair_create(secp256k1_context_sign, keypair, UCharCast(key.data()));
```
Updated.

josibake commented on src/key.cpp:
```diff
@@ -275,10 +275,7 @@ bool CKey::SignSchnorr(const uint256& hash, Span<unsigned char> sig, const uint2
-        if (!secp256k1_keypair_xonly_pub(secp256k1_context_sign, &pubkey, nullptr, &keypair)) return false;
-        unsigned char pubkey_bytes[32];
-        if (!secp256k1_xonly_pubkey_serialize(secp256k1_context_sign, pubkey_bytes, &pubkey)) return false;
+        CPubKey pubkey_bytes = GetPubKey();
```
Thanks! I added a modified version of this as a test. I took out the iterations because creating the keys, signing, etc is an expensive operation and I don't think it adds anything to do multiple rounds every time the tests are ran (i.e., if it this were to fail, it will fail for _all_ keys, not a particular key).

This also means the merkle root isn't be switched, but that's fine since the thing we actually want to test here is that we get the same results using two different methods of getting the public key for the tweak.

josibake commented on src/key.cpp:
```diff
@@ -272,22 +272,19 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
         secp256k1_xonly_pubkey pubkey_verify;
-        ret = secp256k1_keypair_xonly_pub(secp256k1_context_static, &pubkey_verify, nullptr, &keypair);
+        ret = secp256k1_keypair_xonly_pub(secp256k1_context_static, &pubkey_verify, nullptr, keypair);
         ret &= secp256k1_schnorrsig_verify(secp256k1_context_static, sig.data(), hash.begin(), 32, &pubkey_verify);
```
Leaving as is for the reasons mentioned in https://github.com/bitcoin/bitcoin/pull/30051#discussion_r1685528003

paplorinc replied:
One last attempt from my part, but it's not a blocker, would appreciate if you would consider it anyway: in the previous example we've solved this basically by changing the return aggregation (into `ret`) with assertions instead, since we knew that in reality we can't have short-circuiting here. It could make sense to do the same here (if that's the reasoning) and leave a simple return here as well.

itornaza reviewed (approved):
re ACK b8b3a9f18670ec7bf246a57950cdae7e034a264d

Both of you did an amazing job over this weekend and I learned a lot by just watching this thread! Just to re-confirm from my side as well, the aforementioned `KeyPair()` call removal comes without any security implications as we discussed above https://github.com/bitcoin/bitcoin/pull/30051#discussion_r1685779789. However, just for the sake of paranoia I build the commit in debug configuration and run the tests in debug mode using `lldb` and confirmed that `m_keypair` becomes a `nullptr` at the end of `CKey::SignSchnorr()` scope. 

<img width="1034" alt="Screenshot 2024-07-21 at 20 46 38" src="https://github.com/user-attachments/assets/b470587b-0dff-4fb1-9d24-1a42d5691b98">

All unit, functional and extended tests pass as well.

paplorinc commented on src/key.cpp:
```diff
@@ -271,27 +271,9 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
-    memory_cleanse(&keypair, sizeof(keypair));
-    return ret;
+    KeyPair kp = ComputeKeyPair(merkle_root);
+    if (!kp.IsValid()) return false;
+    return kp.SignSchnorr(hash, sig, aux);
```
nit:
```suggestion
    return kp.IsValid() && kp.SignSchnorr(hash, sig, aux);
```

josibake replied:
Updated.

l0rinc reviewed (approved):
utACK b8b3a9f18670ec7bf246a57950cdae7e034a264d

Appreciate your perseverance and creativity @josibake, and your tests @itornaza.

josibake commented:
Update https://github.com/bitcoin/bitcoin/commit/b8b3a9f18670ec7bf246a57950cdae7e034a264d -> https://github.com/bitcoin/bitcoin/commit/56267cf84281789186035f09898cce3d749748ec ([apply-taptweak-method-04](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-04) -> [apply-taptweak-method-05](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-05) ([compare](https://github.com/josibake/bitcoin/compare/apply-taptweak-method-04..josibake:apply-taptweak-method-05))

Realised I unintentionally dropped @theuni as a co-author of the `KeyPair` class. Also applied @paplorinc 's suggestion for `CKey::SignScnorr`. Apologies for making you guys re-ack again, this should be the last one :)

paplorinc commented:
CI failure seems unrelated, please restart or rebase so we can ACK - and don't worry about our reacks :)

> Update https://github.com/bitcoin/bitcoin/commit/b8b3a9f18670ec7bf246a57950cdae7e034a264d -> https://github.com/bitcoin/bitcoin/commit/56267cf84281789186035f09898cce3d749748ec ([apply-taptweak-method-04](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-04) -> [apply-taptweak-method-05](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-05) ([compare](https://github.com/josibake/bitcoin/compare/apply-taptweak-method-04..josibake:apply-taptweak-method-05))

I usually review the changes first via:
<img src="https://github.com/user-attachments/assets/4356aa17-72a2-4198-af6c-8082e102edac">

josibake commented:
Update https://github.com/bitcoin/bitcoin/commit/56267cf84281789186035f09898cce3d749748ec -> https://github.com/bitcoin/bitcoin/commit/e998ca2c53bb8320e102a02a06182f8bc49a4f90 ([apply-taptweak-method-05](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-05) -> [apply-taptweak-method-06](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-06) ([compare](https://github.com/josibake/bitcoin/compare/apply-taptweak-method-05..josibake:apply-taptweak-method-06))

* Fix CI failure (due to moving `test: add key tweak smoke test` up, but without adding `#include secp256k1_keypair`

Not sure about the other CI failures, but if they fail again I'll rebase on master.

> I usually review the changes first via:

Adding the update summary is more for my own book keeping. Also, pushing the individual numbered branches for each change allows reviewers to easily check the diff between force pushes without needing to use github's UI. It also allows reviewers to easily go back and verify old force pushes, whereas I think github deletes the force pushed commits after a certain amount of time?

josibake commented:
Rebased on master https://github.com/bitcoin/bitcoin/commit/e998ca2c53bb8320e102a02a06182f8bc49a4f90 -> https://github.com/bitcoin/bitcoin/commit/9afa2c9e50370b2918377f3f3eac0950a4296ec0 ([apply-taptweak-method-06](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-06) -> [apply-taptweak-method-06](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-06-rebase) ([compare](https://github.com/josibake/bitcoin/compare/apply-taptweak-method-05..josibake:apply-taptweak-method-06-rebase))

josibake commented:
lint CI failure is unrelated and fixed by https://github.com/bitcoin/bitcoin/pull/30499

paplorinc commented:
> lint CI failure is unrelated and fixed by #30499

seems that PR is still failing

paplorinc commented:
utACK 9afa2c9e50370b2918377f3f3eac0950a4296ec0

itornaza commented on src/key.h:
```diff
@@ -235,6 +242,71 @@ struct CExtKey {
+     * for passing a KeyPair object as a secp256k1_keypair and should never be used to access the
+     * underlying keypair bytes directly.
+     */
+    const unsigned char* data() const { return IsValid() ? m_keypair->data() : nullptr; }
```
If there is not a better alternative, I would suggest `secp256k1_keypair()` as a candidate for renaming this `data()` member function to further signify its use through the cast and be called like so:

`reinterpret_cast<const secp256k1_keypair*>(keypair.secp256k1_keypair())`

josibake replied:
(just realised I need to update the description of this PR to be a bit more clear on the motivation)

@sipa - my main motivations for having the `KeyPair` class is to a) encapsulate the taptweak logic so that it can be used outside of signing and b) to be able to pass the `KeyPair` directly to `secp256k1_*` functions expecting a `const secp256k1_keypair *`. For the latter, I think we will need some sort of public function. Here is an example of how this used in #28201:

https://github.com/bitcoin/bitcoin/blob/38067a6ef07d9b5108e75207960bdb7a21c14b70/src/common/bip352.cpp#L206-L221

I do agree a rename would be much better here, perhaps `ReadKeyPairData()/GetKeyPairData()`, or something like @itornaza 's suggestion, `secp256k1_keypair()`?.

Another option is leave the read-only public method out of this PR (since it's not currently used) and add the read-only method in #28201 ?

josibake replied:
I've removed `data()` for now and will add `GetKeyPairData` as a read-only method in #28201

itornaza reviewed (approved):
trACK 9afa2c9e50370b2918377f3f3eac0950a4296ec0

Reviewed the changes in the `key_tests.cpp` since my last review on this PR. Again, all unit and functional including the extended tests pass. Just dropped a name candidate if you consider renaming `KeyPair::data()` and run out of options.

josibake commented on src/test/key_tests.cpp:
```diff
@@ -345,4 +346,31 @@ BOOST_AUTO_TEST_CASE(bip341_test_h)
+    BOOST_CHECK(secp256k1_keypair_xonly_pub(secp256k1_context_sign, &pubkey_old, nullptr, &keypair));
+    unsigned char pubkey_bytes_old[32];
+    BOOST_CHECK(secp256k1_xonly_pubkey_serialize(secp256k1_context_sign, pubkey_bytes_old, &pubkey_old));
+    uint256 tweak_old = XOnlyPubKey(pubkey_bytes_old).ComputeTapTweakHash(merkle_root.IsNull() ? nullptr : &merkle_root);
```
This is how @paplorinc originally proposed the test: https://github.com/bitcoin/bitcoin/pull/30051#discussion_r1685696117

I felt like computing multiple keys in a loop was overkill and removed the merkle root switching since that's not really what's being tested here. The goal here is to sanity check that we get the same public key bytes from GetPubKey() as we would get directly accessing the `secp256k1_keypair` using the `_xonly_pub` and `_serialize`.

So perhaps better to just remove the `IsNull()` check, considering this is being checked in the manner you describe in the `bip340_test_vectors` test?

josibake replied:
I've updated the to remove the `merkle_root.IsNull()` check.

@ismaelsadeeq I kept the `ComputeTapTweakHash` line because they are testing _slightly_ different paths: in the first case, an `XOnlyPubKey` is being constructed from a 32 byte unsigned char array, in the second case it's being created from a `CPubKey` object. I renamed the variables to try to make this more clear in the test.

josibake replied:
Agree the code paths are very similar but I think the same argument as https://github.com/bitcoin/bitcoin/pull/30051#discussion_r1702771127 applies here (admittedly much weaker).

ismaelsadeeq reviewed (commented):
Concept ACK

clean refactor IMO.

paplorinc commented:
To make sure my [benchmarks related post](https://github.com/bitcoin/bitcoin/pull/30051#discussion_r1687854257) don't get lost in resolved comment land, lemme' repost it here:
```patch
diff --git a/src/bench/sign_transaction.cpp b/src/bench/sign_transaction.cpp
--- a/src/bench/sign_transaction.cpp	(revision 8d573611575c3fa66f08407aa9b02f91b29a94c3)
+++ b/src/bench/sign_transaction.cpp	(date 1721731869277)
@@ -12,6 +12,7 @@
 #include <script/script.h>
 #include <script/sign.h>
 #include <uint256.h>
+#include <test/util/random.h>
 #include <util/translation.h>
 
 enum class InputType {
@@ -66,5 +67,33 @@
 static void SignTransactionECDSA(benchmark::Bench& bench)   { SignTransactionSingleInput(bench, InputType::P2WPKH); }
 static void SignTransactionSchnorr(benchmark::Bench& bench) { SignTransactionSingleInput(bench, InputType::P2TR);   }
 
+static void SignSchnorrBenchmark(benchmark::Bench& bench, bool use_null_merkle_root)
+{
+    ECC_Context ecc_context{};
+
+    auto key = GenerateRandomKey();
+    auto msg = InsecureRand256();
+    auto merkle_root = use_null_merkle_root ? uint256() : InsecureRand256();
+    auto aux = InsecureRand256();
+    std::vector<unsigned char> sig(64);
+
+    bench.batch(1).unit("signature").run([&] {
+        bool success = key.SignSchnorr(msg, sig, &merkle_root, aux);
+        assert(success);
+    });
+}
+
+static void SignSchnorrWithMerkleRoot(benchmark::Bench& bench)
+{
+    SignSchnorrBenchmark(bench, /*use_null_merkle_root=*/false);
+}
+
+static void SignSchnorrWithNullMerkleRoot(benchmark::Bench& bench)
+{
+    SignSchnorrBenchmark(bench, /*use_null_merkle_root=*/true);
+}
+
 BENCHMARK(SignTransactionECDSA, benchmark::PriorityLevel::HIGH);
 BENCHMARK(SignTransactionSchnorr, benchmark::PriorityLevel::HIGH);
+BENCHMARK(SignSchnorrWithMerkleRoot, benchmark::PriorityLevel::HIGH);
+BENCHMARK(SignSchnorrWithNullMerkleRoot, benchmark::PriorityLevel::HIGH);
```


which reveals:
before:
|        ns/signature |         signature/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           69,383.89 |           14,412.57 |    0.1% |      1.06 | `SignSchnorrWithMerkleRoot`
|           69,548.30 |           14,378.50 |    0.0% |      1.06 | `SignSchnorrWithNullMerkleRoot`

|        ns/signature |         signature/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           70,189.31 |           14,247.18 |    0.2% |      1.07 | `SignSchnorrWithMerkleRoot`
|           69,464.39 |           14,395.87 |    0.1% |      1.06 | `SignSchnorrWithNullMerkleRoot`


after:
|        ns/signature |         signature/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           83,055.75 |           12,040.11 |    0.2% |      1.08 | `SignSchnorrWithMerkleRoot`
|           80,971.71 |           12,349.99 |    0.1% |      1.08 | `SignSchnorrWithNullMerkleRoot`

|        ns/signature |         signature/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           81,841.15 |           12,218.79 |    0.1% |      1.08 | `SignSchnorrWithMerkleRoot`
|           82,112.31 |           12,178.44 |    0.1% |      1.08 | `SignSchnorrWithNullMerkleRoot`


------

i.e because of the signing code simplifications signing is now ~15% slower.
We could commit the benchmarks and consider a speedup in a separate PR, if this is problematic.

ismaelsadeeq commented on src/key.cpp:
```diff
@@ -272,22 +272,17 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
-        if (!secp256k1_keypair_xonly_tweak_add(secp256k1_context_static, &keypair, tweak.data())) return false;
-    }
-    bool ret = secp256k1_schnorrsig_sign32(secp256k1_context_sign, sig.data(), hash.data(), &keypair, aux.data());
+    KeyPair kp = ComputeKeyPair(merkle_root);
```
nitty-nit:
I think `ConstructKeyPair` might be a better name?

josibake replied:
I prefer to use `Compute*` whenever an actual computation is happening. In this case, calling `ComputeKeyPair` will compute a public key for the private key and (optionally) tweak the newly created key pair.

ismaelsadeeq commented on src/key.cpp:
```diff
@@ -431,6 +420,21 @@ KeyPair::KeyPair(const CKey& key, const uint256* merkle_root)
 
+bool KeyPair::SignSchnorr(const uint256& hash, Span<unsigned char> sig, const uint256& aux) const
+{
+    assert(sig.size() == 64);
```
`m_keypair->data()` can be a `nullptr`, should we return `false` early if it is?

josibake replied:
Good point. Since a KeyPair can be constructed and in an invalid state, I've moved the `if (!IsValid()) return false;` from `CKey::SignSchnorr` to here.

ismaelsadeeq commented on src/test/key_tests.cpp:
```diff
@@ -300,6 +300,13 @@ BOOST_AUTO_TEST_CASE(bip340_test_vectors)
         // Verify those signatures for good measure.
```
Is 9afa2c9e50370b2918377f3f3eac0950a4296ec0 needed, I think this is just a duplicate, because `key.SignSchnorr` is implicitly doing the same thing.

josibake replied:
I don't think there is any harm having a test for both and have a slight preference for keeping it this way only because if someone were to change `CKey::SignSchnorr` in the future (e.g., to not use `KeyPair` or remove the method entirely), `KeyPair` would end up untested.

josibake committed:
bench: add benchmark for signing with a taptweak

Add benchmarks for signing with null and non-null merkle_root arguments.
Null and non-null merkle_root arguments will apply the taptweaks
H_TapTweak(P) and H_TapTweak(P | merkle_root), respectively, to the
private key during signing.

This benchmark is added to verify there are no significant performance
changes after moving the taptweak signing logic in a later commit.

Co-authored-by: l0rinc <pap.lorinc@gmail.com>

josibake committed:
refactor: remove un-tested early returns

Replace early returns in KeyPair::KeyPair() with asserts.

The if statements imply there is an error we are handling, but keypair_xonly_pub
and xonly_pubkey_serialize can only fail if the keypair object is malformed, i.e.,
it was created with a bad secret key. Since we check that the keypair was created
successfully before attempting to extract the public key, using asserts more
accurately documents what we expect here and removes untested branches from the code.

josibake commented:
Update https://github.com/bitcoin/bitcoin/commit/9afa2c9e50370b2918377f3f3eac0950a4296ec0 -> https://github.com/bitcoin/bitcoin/commit/bfb2e6bcef0269141453e43cad56c566c33b9666 ([apply-taptweak-method-06-rebase](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-06-rebase) -> [apply-taptweak-method-07](https://github.com/josibake/bitcoin/tree/apply-taptweak-method-07) ([compare](https://github.com/josibake/bitcoin/compare/apply-taptweak-method-06-rebase..josibake:apply-taptweak-method-07))

* Add `SignSchnorrTapTweakBenchmark` (h/t @paplorinc)
  * Modified the name from @paplorinc 's version and changed it to use `minEpochIterations(100)` instead of `batch(1)`
* Removed public `data()` method. Since this is method is not currently used, it seems better to add this in the Silent Payments sending PR (as `GetKeyPairData()`) where it will be used
* Renamed variables in the schnorr tweak smoke test
* Replaced early returns in the `if merkle_root` block with asserts

l0rinc reviewed (approved):
utACK bfb2e6bcef0269141453e43cad56c566c33b9666

Thanks for your perseverance!

paplorinc commented on src/key.cpp:
```diff
@@ -272,25 +272,17 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
-    bool ret = secp256k1_schnorrsig_sign32(secp256k1_context_sign, sig.data(), hash.data(), &keypair, aux.data());
+    KeyPair kp = ComputeKeyPair(merkle_root);
+    if (!kp.IsValid()) return false;
+    auto keypair = reinterpret_cast<const secp256k1_keypair*>(kp.data());
```
This fails on the commit-by-commit CI task, we still need the getter in this commit - can be removed in the next where we'll access it directly

josibake replied:
Ah, oversight on my part. I meant to combine these two commits in the last push but forgot 😅  The combined commit is still easy enough to review and I think this is more clear than adding a getter and then removing it in the next commit.

ismaelsadeeq reviewed (approved):
Code review ACK ec973dd19719541dbcd6f3a6facf6f5dd7cf439c

paplorinc commented:
ACK ec973dd19719541dbcd6f3a6facf6f5dd7cf439c - will happily reack if you decide to apply @ismaelsadeeq's suggestions

josibake commented on src/test/key_tests.cpp:
```diff
@@ -345,4 +346,31 @@ BOOST_AUTO_TEST_CASE(bip341_test_h)
+BOOST_AUTO_TEST_CASE(key_schnorr_tweak_smoke_test)
+{
+    // Sanity check to ensure we get the same tweak using CPubKey vs secp256k1 functions
+    secp256k1_context* secp256k1_context_sign = secp256k1_context_create(SECP256K1_CONTEXT_SIGN);
```
Ah, nice! Good catch. Will change if I end up retouching.

paplorinc replied:
If you do that, please change other usages as well

josibake replied:
Yeah, did a quick grep and it looks there is at least one other place in the fuzz tests where `_SIGN` is used, so this one is probably best as a follow up PR to replace all instances with `_NONE` in the codebase and perhaps add a mention to the developer notes.

josibake commented on src/key.cpp:
```diff
@@ -425,16 +406,39 @@ void CExtKey::Decode(const unsigned char code[BIP32_EXTKEY_SIZE]) {
+        secp256k1_xonly_pubkey pubkey;
+        if (!secp256k1_keypair_xonly_pub(secp256k1_context_sign, &pubkey, nullptr, keypair)) return;
+        unsigned char pubkey_bytes[32];
+        if (!secp256k1_xonly_pubkey_serialize(secp256k1_context_sign, pubkey_bytes, &pubkey)) return;
```
Not sure if this is still applicable after the change in the last commit to replace the if branches with asserts, but will take a look and updated if I end up needing to retouch.

itornaza reviewed (approved):
trACK ec973dd19719541dbcd6f3a6facf6f5dd7cf439c

Rebuilt the commit, run all unit, functional and extended tests and all of them pass.

theStack reviewed (approved):
Code-review ACK ec973dd19719541dbcd6f3a6facf6f5dd7cf439c

Left a few non-critical nits below, feel free to ignore.

paplorinc commented on src/test/key_tests.cpp:
```diff
@@ -345,4 +346,31 @@ BOOST_AUTO_TEST_CASE(bip341_test_h)
+    BOOST_CHECK(secp256k1_keypair_xonly_pub(secp256k1_context_sign, &xonly_pubkey, nullptr, &keypair));
+    unsigned char xonly_bytes[32];
+    BOOST_CHECK(secp256k1_xonly_pubkey_serialize(secp256k1_context_sign, xonly_bytes, &xonly_pubkey));
+    uint256 tweak_old = XOnlyPubKey(xonly_bytes).ComputeTapTweakHash(&merkle_root);
+
+    // CPubKey
+    CPubKey pubkey = key.GetPubKey();
+    uint256 tweak_new = XOnlyPubKey(pubkey).ComputeTapTweakHash(&merkle_root);
```
This was the result of a change that was reverted since, where it was important to test the exact before/after structure.

paplorinc commented on src/key.cpp:
```diff
@@ -414,9 +414,9 @@ KeyPair::KeyPair(const CKey& key, const uint256* merkle_root)
-        if (!secp256k1_keypair_xonly_pub(secp256k1_context_sign, &pubkey, nullptr, keypair)) return;
         unsigned char pubkey_bytes[32];
-        if (!secp256k1_xonly_pubkey_serialize(secp256k1_context_sign, pubkey_bytes, &pubkey)) return;
+        assert(secp256k1_keypair_xonly_pub(secp256k1_context_sign, &pubkey, nullptr, keypair));
+        assert(secp256k1_xonly_pubkey_serialize(secp256k1_context_sign, pubkey_bytes, &pubkey));
```
We've had something like that before: https://github.com/bitcoin/bitcoin/pull/30051#discussion_r1681077911

But it was ambiguous, as to whether it can fail or not during execution as we can see here:
https://github.com/hebasto/bitcoin/blob/master/src/pubkey.cpp#L345-L350 or https://github.com/hebasto/bitcoin/blob/master/src/secp256k1/src/secp256k1.c#L685

But @josibake argued that they cannot fail at that point - if they do it's a bug, but we didn't want to ignore the return values either. So we ended up asserting, like we did in [GetPubKey](https://github.com/hebasto/bitcoin/blob/master/src/key.cpp#L188) already.

Shuhaib07 commented:
Merge and get approved

Shuhaib07 reviewed (commented):
Okay

hebasto commented:
The buildsystem related stuff has been ported to the CMake-based buildsystem in https://github.com/hebasto/bitcoin/pull/317.

Shuhaib07 reviewed (commented):
Approve
